synapse/replication/tcp/client.py

```diff
@@ -219,6 +219,21 @@ async def on_rdata(
                     membership=row.data.membership,
                 )
 
+                # If this event is a join, make a note of it so we have an accurate
+                # cross-worker room rate limit.
+                # TODO: Erik said we should exclude rows that came from ex_outliers
+                #  here, but I don't see how we can determine that. I guess we could
+                #  add a flag to row.data?
+                if (
+                    row.data.type == EventTypes.Member
+                    and row.data.membership == Membership.JOIN
+                    and not row.data.outlier
```

Comment:
Purely for my own edification, why are events that are outliers excluded from being counted?

Reply:
Oh god, erm. @erikjohnston explained this to me before. Very roughly speaking: an outlier is an event which we have seen but haven't fully processed or integrated into our view of the room yet. E.g. it might be something very old that we're only just learning about. In particular, we don't send outliers down to clients, so it doesn't seem appropriate to include them in the ratelimit tracking.

An event can go from being an outlier to not an outlier and I'm not quite sure what the right thing to do is here. I've chosen to ignore it and see if this implementation is good enough (TM).

Reply:
A specific example of this is: when a server joins an existing room it will start with the join event and the state at that join event. Those events are then streamed down replication: the join event is a non-outlier, but all the state are outliers.